File: leetcode/0040/combinationSum2.py
from typing import List
def combinationSum2(candidates: List[int], target: int) -> List[List[int]]:
    def dfs(begin, cur, target):
        if target == 0:
            res.append(cur[:])
            return

        for index in range(begin, size):
            if candidates[index] > target:
                break

            if index > begin and candidates[index - 1] == candidates[index]:
                continue

            cur.append(candidates[index])
            dfs(index + 1, cur, target - candidates[index])
            cur.pop()

    size = len(candidates)
    if size == 0:
        return []

    candidates.sort()
    res = []
    dfs(0, [], target)
    return res

print(combinationSum2([10,1,2,7,6,1,5], 8))

File: leetcode/0040/combinationSum2.swift
var n = 0
var ans = [[Int]]()
var sortedCandidates = [Int]()
func combinationSum2(_ candidates: [Int], _ target: Int) -> [[Int]] {
    n = candidates.count
    ans = [[Int]]()
    sortedCandidates = candidates.sorted()
    
    dfs(target, [], 0)
    
    return ans
}

func dfs(_ target: Int, _ cur: [Int], _ begin: Int) {
    var cur = cur
    if target == 0 {
        ans.append(cur)
    } else {
        for i in begin..<n {
            if sortedCandidates[i] > target {
                break
            }
            if i > begin && sortedCandidates[i - 1] == sortedCandidates[i] {
                continue
            }
            cur.append(sortedCandidates[i])
            dfs(target-sortedCandidates[i], cur, i+1)
            cur.popLast()
        }
    }
}

print(combinationSum2([10,1,2,7,6,1,5], 8))
